src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -0,0 +1,101 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package picard.sam.markduplicates;
+
+import htsjdk.samtools.DuplicateSet;
+import htsjdk.samtools.DuplicateSetIterator;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.util.CloseableIterator;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.counting;
+
+public class UmiAwareDuplicateSetIterator extends UmiGraph implements CloseableIterator<DuplicateSet> {
+    private final DuplicateSetIterator wrappedIterator;
+    private Iterator<DuplicateSet> nextSetsIterator;
+    private final int editDistanceToJoin;
+    private final boolean addInferredUmi;
+    private final String umiTag;
+    private final String inferredUmiTag;
+
+    public UmiAwareDuplicateSetIterator(final DuplicateSetIterator wrappedIterator, final int editDistanceToJoin, final boolean addInferredUmi, final String umiTag, final String inferredUmiTag) {
+        this.wrappedIterator = wrappedIterator;
+        this.editDistanceToJoin = editDistanceToJoin;
+        this.addInferredUmi = addInferredUmi;
+        this.umiTag = umiTag;
+        this.inferredUmiTag = inferredUmiTag;
+        nextSetsIterator = Collections.emptyIterator();
+    }
+
+    @Override
+    public void close() {
+        wrappedIterator.close();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return nextSetsIterator.hasNext() || wrappedIterator.hasNext();
+    }
+
+    @Override
+    public DuplicateSet next() {
+        if(!nextSetsIterator.hasNext())
+            process(wrappedIterator.next());
+
+        return nextSetsIterator.next();
+    }
+
+    // Takes a duplicate set and breaks it up into possible smaller sets according to the UMI,
+    // and updates nextSetsIterator to be an iterator on that set of DuplicateSets.
+    private void process(final DuplicateSet set) {
+
+        final List<SAMRecord> records = set.getRecords();
+
+        // If any records are missing the UMI_TAG proceed as if there were no UMIs
+        // and return nextSetsIterator without breaking it up into smaller sets.
```

Comment:
@nh13 I'm not sure why UMIs might be missing.  If they are, what do you think would be the appropriate behavior, throw an exception?

I've recently modified the code to treat a non-existent UMI to be treated as if it were it's own unique identifier.